packages/node_modules/pouchdb-mapreduce/src/index.js

```diff
@@ -505,48 +504,59 @@ function updateViewInQueue(view) {
       style: 'all_docs',
       since: currentSeq,
       limit: CHANGES_BATCH_SIZE
-    }).then(function (response) {
-      var results = response.results;
-      if (!results.length) {
-        return;
-      }
-      var docIdsToChangesAndEmits = new Map();
-      for (var i = 0, l = results.length; i < l; i++) {
-        var change = results[i];
-        if (change.doc._id[0] !== '_') {
-          mapResults = [];
-          doc = change.doc;
-
-          if (!doc._deleted) {
-            tryMap(view.sourceDB, mapFun, doc);
-          }
-          mapResults.sort(sortByKeyThenValue);
-
-          var indexableKeysToKeyValues = {};
-          var lastKey;
-          for (var j = 0, jl = mapResults.length; j < jl; j++) {
-            var obj = mapResults[j];
-            var complexKey = [obj.key, obj.id];
-            if (collate(obj.key, lastKey) === 0) {
-              complexKey.push(j); // dup key+id, so make it unique
-            }
-            var indexableKey = toIndexableString(complexKey);
-            indexableKeysToKeyValues[indexableKey] = obj;
-            lastKey = obj.key;
-          }
-          docIdsToChangesAndEmits.set(change.doc._id, {
-            indexableKeysToKeyValues: indexableKeysToKeyValues,
-            changes: change.changes
-          });
+    }).then(processBatch);
+  }
+
+  function processBatch(response) {
+    var results = response.results;
+    if (!results.length) {
+      return;
+    }
+    var docIdsToChangesAndEmits = createDocIdsToChangesAndEmits(results);
+    queue.add(processChange(docIdsToChangesAndEmits, currentSeq));
+    if (results.length < CHANGES_BATCH_SIZE) {
+      return;
+    }
+    return processNextBatch();
+  }
+
+  function createDocIdsToChangesAndEmits(results) {
+    var docIdsToChangesAndEmits = new Map();
+    for (var i = 0, len = results.length; i < len; i++) {
+      var change = results[i];
+      if (change.doc._id[0] !== '_') {
+        mapResults = [];
+        doc = change.doc;
+
+        if (!doc._deleted) {
+          tryMap(view.sourceDB, mapFun, doc);
         }
-        currentSeq = change.seq;
+        mapResults.sort(sortByKeyThenValue);
+
+        var indexableKeysToKeyValues = createIndexableKeysToKeyValues(mapResults);
+        docIdsToChangesAndEmits.set(change.doc._id, [
+          indexableKeysToKeyValues,
+          change.changes
+        ]);
       }
-      queue.add(processChange(docIdsToChangesAndEmits, currentSeq));
-      if (results.length < CHANGES_BATCH_SIZE) {
-        return;
+      currentSeq = change.seq;
+    }
+    return docIdsToChangesAndEmits;
+  }
+
+  function createIndexableKeysToKeyValues(mapResults) {
+    var indexableKeysToKeyValues = new Map();
+    var lastKey;
```

Comment:
it turned out I had forgotten about JavaScript variable hoisting, and this `lastKey` was non-undefined between doc comparisons, meaning that we were doing needless comparisons here. it's only necessary to uniqify the indexable key for an emission if it's 1) the same document 2) emitting the exact same value. So by moving this to a separate function, we re-initialize `lastKey` to undefined every time, and furthermore by checking `i > 0`, this is what led to those old `pouchdb-collate` code paths not being hit anymore.

Furthermore this means that from now on our indexable keys should be quite a bit simpler now; no need to add the `i` value unless we truly are emitting the exact same key for the exact same document.